sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidRels.java

```diff
@@ -59,12 +59,14 @@ public static boolean isScanOrMapping(final DruidRel<?> druidRel, final boolean
   /**
    * Check if a druidRel is a simple table scan or a scan + projection.
    *
-   * @param druidRel  the rel to check
-   * @param canBeJoin consider a 'join' that doesn't do anything fancy to be a scan-or-mapping too.
+   * @param druidRel         the rel to check
+   * @param canBeJoinOrUnion consider a {@link DruidJoinQueryRel} or {@link DruidUnionDataSourceRel} as possible
+   *                         scans-and-mappings too.
    */
-  private static boolean isScanOrProject(final DruidRel<?> druidRel, final boolean canBeJoin)
+  private static boolean isScanOrProject(final DruidRel<?> druidRel, final boolean canBeJoinOrUnion)
   {
-    if (druidRel instanceof DruidQueryRel || (canBeJoin && druidRel instanceof DruidJoinQueryRel)) {
+    if (druidRel instanceof DruidQueryRel || (canBeJoinOrUnion && (druidRel instanceof DruidJoinQueryRel
```

Comment:
nit: I wonder if the `instanceof` checks here are something that should be pushed into a couple of methods on `DruidRel` (one to check if is scan/projection one to check if is 'external' (or is there a better word?) scan/projection for join/union)

Reply:
You might be right. I have a sneaking suspicion that we'll want to redesign the DruidRel interface at some point. I was thinking we could rethink it when we create a native datasource that represents a union of queries, which would allow us to merge the two UNION ALL implementations in SQL, which would allow a single-SQL-query-to-single-native-query mapping. That would simplify a lot of stuff and would be a good catalyst for rethinking the interface.